File: frontend/src/lib/api/webproxy.ts
import { apiClient } from "./client";

// ============================================================================
// Types
// ============================================================================

export interface WebProxyLdap {
  server?: string | null;
  base_dn?: string | null;
  bind_dn?: string | null;
  filter_expression?: string | null;
  password?: string | null;
  username_attribute?: string | null;
  port?: number | null;
  version?: string | null;
  persistent_connection: boolean;
  use_ssl: boolean;
}

export interface WebProxyAuthentication {
  children?: number | null;
  credentials_ttl?: number | null;
  method?: string | null;
  realm?: string | null;
  ldap: WebProxyLdap;
}

export interface CachePeer {
  name: string;
  address?: string | null;
  http_port?: number | null;
  icp_port?: number | null;
  options?: string | null;
  type?: string | null;
}

export interface ListenAddress {
  address: string;
  port?: number | null;
  disable_transparent: boolean;
}

export interface SquidGuardRule {
  number: string;
  allow_categories: string[];
  block_categories: string[];
  log: string[];
  local_block: string[];
  local_block_keyword: string[];
  local_block_url: string[];
  local_ok: string[];
  local_ok_url: string[];
  allow_ipaddr_url: boolean;
  enable_safe_search: boolean;
  default_action?: string | null;
  redirect_url?: string | null;
  source_group?: string | null;
  time_period?: string | null;
}

export interface SquidGuardSourceGroup {
  name: string;
  address: string[];
  domain: string[];
  ldap_ip_search: string[];
  ldap_user_search: string[];
  description?: string | null;
  user?: string | null;
}

export interface TimePeriodDay {
  day: string;
  time?: string | null;
}

export interface SquidGuardTimePeriod {
  name: string;
  description?: string | null;
  days: TimePeriodDay[];
}

export interface SquidGuard {
  allow_categories: string[];
  block_categories: string[];
  log: string[];
  local_block: string[];
  local_block_keyword: string[];
  local_block_url: string[];
  local_ok: string[];
  local_ok_url: string[];
  allow_ipaddr_url: boolean;
  enable_safe_search: boolean;
  default_action?: string | null;
  redirect_url?: string | null;
  auto_update_hour?: number | null;
  rules: SquidGuardRule[];
  source_groups: SquidGuardSourceGroup[];
  time_periods: SquidGuardTimePeriod[];
}

export interface UrlFiltering {
  disable: boolean;
  squidguard: SquidGuard;
}

export interface WebProxyConfig {
  append_domain?: string | null;
  cache_size?: number | null;
  default_port?: number | null;
  maximum_object_size?: number | null;
  mem_cache_size?: number | null;
  minimum_object_size?: number | null;
  outgoing_address?: string | null;
  reply_body_max_size?: number | null;
  disable_access_log: boolean;
  domain_block: string[];
  domain_noncache: string[];
  reply_block_mime: string[];
  safe_ports: string[];
  ssl_safe_ports: string[];
  authentication: WebProxyAuthentication;
  cache_peers: CachePeer[];
  listen_addresses: ListenAddress[];
  url_filtering: UrlFiltering;
}

export interface WebProxyCapabilities {
  version: string;
  features: {
    global_settings: { supported: boolean; description: string };
    authentication: { supported: boolean; description: string };
    cache_peer: { supported: boolean; description: string };
    listen_address: { supported: boolean; description: string };
    url_filtering: { supported: boolean; description: string };
  };
  options: {
    cache_peer_type: string[];
    default_action: string[];
    ldap_version: string[];
    auth_method: string[];
    time_period_days: string[];
    reply_block_mime: string[];
  };
  version_info: { is_1_4: boolean; is_1_5: boolean };
}

export interface BatchOperation {
  op: string;
  value?: string | null;
}

export interface VyOSResponse {
  success: boolean;
  data?: Record<string, unknown> | null;
  error?: string | null;
}

// ============================================================================
// Service
// ============================================================================

class WebProxyService {
  async getCapabilities(): Promise<WebProxyCapabilities> {
    return apiClient.get<WebProxyCapabilities>("/vyos/webproxy/capabilities");
  }

  async getConfig(refresh = false): Promise<WebProxyConfig> {
    return apiClient.get<WebProxyConfig>("/vyos/webproxy/config", {
      refresh: refresh.toString(),
    });
  }

  private async batch(operations: BatchOperation[]): Promise<VyOSResponse> {
    const result = await apiClient.post<VyOSResponse>("/vyos/webproxy/batch", {
      operations,
    });
    if (!result.success) {
      throw new Error(result.error || "Operation failed");
    }
    return result;
  }

  // ----- Global settings + lists ------------------------------------------

  async saveSettings(config: WebProxyConfig): Promise<VyOSResponse> {
    const ops: BatchOperation[] = [];

    const scalar = (op: string, value: number | string | null | undefined) => {
      ops.push({ op: `delete_${op}` });
      if (value !== null && value !== undefined && String(value) !== "") {
        ops.push({ op: `set_${op}`, value: String(value) });
      }
    };

    scalar("append_domain", config.append_domain);
    scalar("cache_size", config.cache_size);
    scalar("default_port", config.default_port);
    scalar("maximum_object_size", config.maximum_object_size);
    scalar("mem_cache_size", config.mem_cache_size);
    scalar("minimum_object_size", config.minimum_object_size);
    scalar("outgoing_address", config.outgoing_address);
    scalar("reply_body_max_size", config.reply_body_max_size);

    ops.push({ op: "delete_disable_access_log" });
    if (config.disable_access_log) ops.push({ op: "set_disable_access_log" });

    const list = (
      addOp: string,
      delAllOp: string,
      values: string[]
    ) => {
      ops.push({ op: delAllOp });
      for (const v of values) ops.push({ op: addOp, value: v });
    };

    list("add_domain_block", "delete_domain_block_all", config.domain_block);
    list("add_domain_noncache", "delete_domain_noncache_all", config.domain_noncache);
    list("add_reply_block_mime", "delete_reply_block_mime_all", config.reply_block_mime);
    list("add_safe_port", "delete_safe_port_all", config.safe_ports);
    list("add_ssl_safe_port", "delete_ssl_safe_port_all", config.ssl_safe_ports);

    return this.batch(ops);
  }

  // ----- Authentication ----------------------------------------------------

  async saveAuthentication(auth: WebProxyAuthentication): Promise<VyOSResponse> {
    const ops: BatchOperation[] = [];
    // Replace the whole authentication subtree for a clean state.
    ops.push({ op: "delete_authentication" });

    const set = (op: string, value: number | string | null | undefined) => {
      if (value !== null && value !== undefined && String(value) !== "") {
        ops.push({ op, value: String(value) });
      }
    };

    set("set_authentication_children", auth.children);
    set("set_authentication_credentials_ttl", auth.credentials_ttl);
    set("set_authentication_method", auth.method);
    set("set_authentication_realm", auth.realm);

    const ldap = auth.ldap;
    set("set_ldap_server", ldap.server);
    set("set_ldap_base_dn", ldap.base_dn);
    set("set_ldap_bind_dn", ldap.bind_dn);
    set("set_ldap_filter_expression", ldap.filter_expression);
    set("set_ldap_password", ldap.password);
    set("set_ldap_username_attribute", ldap.username_attribute);
    set("set_ldap_port", ldap.port);
    set("set_ldap_version", ldap.version);
    if (ldap.persistent_connection) ops.push({ op: "set_ldap_persistent_connection" });
    if (ldap.use_ssl) ops.push({ op: "set_ldap_use_ssl" });

    return this.batch(ops);
  }

  async clearAuthentication(): Promise<VyOSResponse> {
    return this.batch([{ op: "delete_authentication" }]);
  }

  // ----- Cache peers -------------------------------------------------------

  async saveCachePeer(peer: CachePeer, isEdit: boolean): Promise<VyOSResponse> {
    const ops: BatchOperation[] = [];
    const name = peer.name;
    if (isEdit) ops.push({ op: "delete_cache_peer", value: name });
    ops.push({ op: "set_cache_peer", value: name });

    const set = (op: string, value: number | string | null | undefined) => {
      if (value !== null && value !== undefined && String(value) !== "") {
        ops.push({ op, value: `${name},${value}` });
      }
    };
    set("set_cache_peer_address", peer.address);
    set("set_cache_peer_http_port", peer.http_port);
    set("set_cache_peer_icp_port", peer.icp_port);
    set("set_cache_peer_options", peer.options);
    set("set_cache_peer_type", peer.type);

    return this.batch(ops);
  }

  async deleteCachePeer(name: string): Promise<VyOSResponse> {
    return this.batch([{ op: "delete_cache_peer", value: name }]);
  }

  // ----- Listen addresses --------------------------------------------------

  async saveListenAddress(addr: ListenAddress, isEdit: boolean): Promise<VyOSResponse> {
    const ops: BatchOperation[] = [];
    const address = addr.address;
    if (isEdit) ops.push({ op: "delete_listen_address", value: address });
    ops.push({ op: "set_listen_address", value: address });

    if (addr.port !== null && addr.port !== undefined && String(addr.port) !== "") {
      ops.push({ op: "set_listen_address_port", value: `${address},${addr.port}` });
    }
    if (addr.disable_transparent) {
      ops.push({ op: "set_listen_address_disable_transparent", value: address });
    }
    return this.batch(ops);
  }

  async deleteListenAddress(address: string): Promise<VyOSResponse> {
    return this.batch([{ op: "delete_listen_address", value: address }]);
  }

  // ----- URL filtering -----------------------------------------------------

  async setUrlFilteringDisabled(disabled: boolean): Promise<VyOSResponse> {
    return this.batch([
      { op: disabled ? "set_url_filtering_disable" : "delete_url_filtering_disable" },
    ]);
  }

  async saveSquidGuardGlobal(sg: SquidGuard): Promise<VyOSResponse> {
    const ops: BatchOperation[] = [];

    const list = (addOp: string, delAllOp: string, values: string[]) => {
      ops.push({ op: delAllOp });
      for (const v of values) ops.push({ op: addOp, value: v });
    };
    list("add_squidguard_allow_category", "delete_squidguard_allow_category_all", sg.allow_categories);
    list("add_squidguard_block_category", "delete_squidguard_block_category_all", sg.block_categories);
    list("add_squidguard_log", "delete_squidguard_log_all", sg.log);
    list("add_squidguard_local_block", "delete_squidguard_local_block_all", sg.local_block);
    list("add_squidguard_local_block_keyword", "delete_squidguard_local_block_keyword_all", sg.local_block_keyword);
    list("add_squidguard_local_block_url", "delete_squidguard_local_block_url_all", sg.local_block_url);
    list("add_squidguard_local_ok", "delete_squidguard_local_ok_all", sg.local_ok);
    list("add_squidguard_local_ok_url", "delete_squidguard_local_ok_url_all", sg.local_ok_url);

    ops.push({ op: "delete_squidguard_allow_ipaddr_url" });
    if (sg.allow_ipaddr_url) ops.push({ op: "set_squidguard_allow_ipaddr_url" });
    ops.push({ op: "delete_squidguard_enable_safe_search" });
    if (sg.enable_safe_search) ops.push({ op: "set_squidguard_enable_safe_search" });

    ops.push({ op: "delete_squidguard_default_action" });
    if (sg.default_action) ops.push({ op: "set_squidguard_default_action", value: sg.default_action });
    ops.push({ op: "delete_squidguard_redirect_url" });
    if (sg.redirect_url) ops.push({ op: "set_squidguard_redirect_url", value: sg.redirect_url });
    ops.push({ op: "delete_squidguard_auto_update" });
    if (sg.auto_update_hour !== null && sg.auto_update_hour !== undefined) {
      ops.push({ op: "set_squidguard_auto_update_hour", value: String(sg.auto_update_hour) });
    }

    return this.batch(ops);
  }

  // ----- squidGuard rules --------------------------------------------------

  async saveRule(rule: SquidGuardRule, isEdit: boolean): Promise<VyOSResponse> {
    const ops: BatchOperation[] = [];
    const n = rule.number;
    if (isEdit) ops.push({ op: "delete_squidguard_rule", value: n });
    ops.push({ op: "set_squidguard_rule", value: n });

    const list = (addOp: string, values: string[]) => {
      for (const v of values) ops.push({ op: addOp, value: `${n},${v}` });
    };
    list("add_squidguard_rule_allow_category", rule.allow_categories);
    list("add_squidguard_rule_block_category", rule.block_categories);
    list("add_squidguard_rule_log", rule.log);
    list("add_squidguard_rule_local_block", rule.local_block);
    list("add_squidguard_rule_local_block_keyword", rule.local_block_keyword);
    list("add_squidguard_rule_local_block_url", rule.local_block_url);
    list("add_squidguard_rule_local_ok", rule.local_ok);
    list("add_squidguard_rule_local_ok_url", rule.local_ok_url);

    if (rule.allow_ipaddr_url) ops.push({ op: "set_squidguard_rule_allow_ipaddr_url", value: n });
    if (rule.enable_safe_search) ops.push({ op: "set_squidguard_rule_enable_safe_search", value: n });

    const set = (op: string, value: string | null | undefined) => {
      if (value) ops.push({ op, value: `${n},${value}` });
    };
    set("set_squidguard_rule_default_action", rule.default_action);
    set("set_squidguard_rule_redirect_url", rule.redirect_url);
    set("set_squidguard_rule_source_group", rule.source_group);
    set("set_squidguard_rule_time_period", rule.time_period);

    return this.batch(ops);
  }

  async deleteRule(number: string): Promise<VyOSResponse> {
    return this.batch([{ op: "delete_squidguard_rule", value: number }]);
  }

  // ----- squidGuard source groups -----------------------------------------

  async saveSourceGroup(group: SquidGuardSourceGroup, isEdit: boolean): Promise<VyOSResponse> {
    const ops: BatchOperation[] = [];
    const name = group.name;
    if (isEdit) ops.push({ op: "delete_squidguard_source_group", value: name });
    ops.push({ op: "set_squidguard_source_group", value: name });

    const list = (addOp: string, values: string[]) => {
      for (const v of values) ops.push({ op: addOp, value: `${name},${v}` });
    };
    list("add_squidguard_source_group_address", group.address);
    list("add_squidguard_source_group_domain", group.domain);
    list("add_squidguard_source_group_ldap_ip_search", group.ldap_ip_search);
    list("add_squidguard_source_group_ldap_user_search", group.ldap_user_search);

    if (group.description) {
      ops.push({ op: "set_squidguard_source_group_description", value: `${name},${group.description}` });
    }
    if (group.user) {
      ops.push({ op: "set_squidguard_source_group_user", value: `${name},${group.user}` });
    }
    return this.batch(ops);
  }

  async deleteSourceGroup(name: string): Promise<VyOSResponse> {
    return this.batch([{ op: "delete_squidguard_source_group", value: name }]);
  }

  // ----- squidGuard time periods ------------------------------------------

  async saveTimePeriod(period: SquidGuardTimePeriod, isEdit: boolean): Promise<VyOSResponse> {
    const ops: BatchOperation[] = [];
    const name = period.name;
    if (isEdit) ops.push({ op: "delete_squidguard_time_period", value: name });
    ops.push({ op: "set_squidguard_time_period", value: name });

    if (period.description) {
      ops.push({ op: "set_squidguard_time_period_description", value: `${name},${period.description}` });
    }
    for (const d of period.days) {
      if (d.time) {
        ops.push({ op: "set_squidguard_time_period_day_time", value: `${name},${d.day},${d.time}` });
      }
    }
    return this.batch(ops);
  }

  async deleteTimePeriod(name: string): Promise<VyOSResponse> {
    return this.batch([{ op: "delete_squidguard_time_period", value: name }]);
  }
}

export const webProxyService = new WebProxyService();

File: backend/routers/webproxy/webproxy.py
"""
Web Proxy (Squid) Router

API endpoints for managing the VyOS webproxy service (service webproxy).

Three endpoints:
  GET  /vyos/webproxy/capabilities  — version-aware feature flags + option lists
  GET  /vyos/webproxy/config        — normalized webproxy configuration
  POST /vyos/webproxy/batch         — atomic set/delete operations

The webproxy CLI is identical on VyOS 1.4 and 1.5.
"""

from fastapi import APIRouter, HTTPException, Request
from starlette.concurrency import run_in_threadpool
from pydantic import BaseModel, Field
from typing import List, Dict, Optional, Any
from session_vyos_service import get_session_vyos_service
from vyos_builders.webproxy import WebProxyBatchBuilder
from fastapi_permissions import require_read_permission, require_write_permission
from rbac_permissions import FeatureGroup
import inspect
import logging

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/vyos/webproxy", tags=["webproxy"])


# ============================================================================
# Pydantic Models
# ============================================================================


class WebProxyLdap(BaseModel):
    server: Optional[str] = None
    base_dn: Optional[str] = None
    bind_dn: Optional[str] = None
    filter_expression: Optional[str] = None
    password: Optional[str] = None
    username_attribute: Optional[str] = None
    port: Optional[int] = None
    version: Optional[str] = None
    persistent_connection: bool = False
    use_ssl: bool = False


class WebProxyAuthentication(BaseModel):
    children: Optional[int] = None
    credentials_ttl: Optional[int] = None
    method: Optional[str] = None
    realm: Optional[str] = None
    ldap: WebProxyLdap = Field(default_factory=WebProxyLdap)


class CachePeer(BaseModel):
    name: str
    address: Optional[str] = None
    http_port: Optional[int] = None
    icp_port: Optional[int] = None
    options: Optional[str] = None
    type: Optional[str] = None


class ListenAddress(BaseModel):
    address: str
    port: Optional[int] = None
    disable_transparent: bool = False


class SquidGuardRule(BaseModel):
    number: str
    allow_categories: List[str] = []
    block_categories: List[str] = []
    log: List[str] = []
    local_block: List[str] = []
    local_block_keyword: List[str] = []
    local_block_url: List[str] = []
    local_ok: List[str] = []
    local_ok_url: List[str] = []
    allow_ipaddr_url: bool = False
    enable_safe_search: bool = False
    default_action: Optional[str] = None
    redirect_url: Optional[str] = None
    source_group: Optional[str] = None
    time_period: Optional[str] = None


class SquidGuardSourceGroup(BaseModel):
    name: str
    address: List[str] = []
    domain: List[str] = []
    ldap_ip_search: List[str] = []
    ldap_user_search: List[str] = []
    description: Optional[str] = None
    user: Optional[str] = None


class TimePeriodDay(BaseModel):
    day: str
    time: Optional[str] = None


class SquidGuardTimePeriod(BaseModel):
    name: str
    description: Optional[str] = None
    days: List[TimePeriodDay] = []


class SquidGuard(BaseModel):
    allow_categories: List[str] = []
    block_categories: List[str] = []
    log: List[str] = []
    local_block: List[str] = []
    local_block_keyword: List[str] = []
    local_block_url: List[str] = []
    local_ok: List[str] = []
    local_ok_url: List[str] = []
    allow_ipaddr_url: bool = False
    enable_safe_search: bool = False
    default_action: Optional[str] = None
    redirect_url: Optional[str] = None
    auto_update_hour: Optional[int] = None
    rules: List[SquidGuardRule] = []
    source_groups: List[SquidGuardSourceGroup] = []
    time_periods: List[SquidGuardTimePeriod] = []


class UrlFiltering(BaseModel):
    disable: bool = False
    squidguard: SquidGuard = Field(default_factory=SquidGuard)


class WebProxyConfig(BaseModel):
    append_domain: Optional[str] = None
    cache_size: Optional[int] = None
    default_port: Optional[int] = None
    maximum_object_size: Optional[int] = None
    mem_cache_size: Optional[int] = None
    minimum_object_size: Optional[int] = None
    outgoing_address: Optional[str] = None
    reply_body_max_size: Optional[int] = None
    disable_access_log: bool = False
    domain_block: List[str] = []
    domain_noncache: List[str] = []
    reply_block_mime: List[str] = []
    safe_ports: List[str] = []
    ssl_safe_ports: List[str] = []
    authentication: WebProxyAuthentication = Field(default_factory=WebProxyAuthentication)
    cache_peers: List[CachePeer] = []
    listen_addresses: List[ListenAddress] = []
    url_filtering: UrlFiltering = Field(default_factory=UrlFiltering)


class WebProxyBatchOperation(BaseModel):
    op: str = Field(..., description="Builder method name")
    value: Optional[str] = Field(
        None,
        description=(
            "Argument(s) for the operation. "
            "Single-arg methods: plain string. "
            "Multi-arg methods: comma-separated."
        ),
    )


class WebProxyBatchRequest(BaseModel):
    operations: List[WebProxyBatchOperation]


class VyOSResponse(BaseModel):
    success: bool
    data: Optional[Dict[str, Any]] = None
    error: Optional[str] = None


# ============================================================================
# Internal builder method denylist
# ============================================================================

_INTERNAL_BUILDER_METHODS = frozenset({
    "add_set", "add_delete", "get_operations", "is_empty",
    "get_capabilities", "mappers", "version", "_operations", "m",
})


# ============================================================================
# Endpoint 1: Capabilities
# ============================================================================


@router.get("/capabilities")
async def get_webproxy_capabilities(request: Request):
    """Return webproxy feature capabilities based on the VyOS version."""
    await require_read_permission(request, FeatureGroup.WEBPROXY)
    try:
        service = get_session_vyos_service(request)
        builder = WebProxyBatchBuilder(version=service.get_version())
        caps = builder.get_capabilities()
        if hasattr(request.state, "instance") and request.state.instance:
            caps["instance_name"] = request.state.instance.get("name")
            caps["instance_id"] = request.state.instance.get("id")
        return caps
    except Exception:
        logger.exception("Unhandled error in get_webproxy_capabilities")
        raise HTTPException(status_code=500, detail="Internal server error")


# ============================================================================
# Endpoint 2: Config
# ============================================================================


@router.get("/config", response_model=WebProxyConfig)
async def get_webproxy_config(http_request: Request, refresh: bool = False):
    """Return the full webproxy configuration in a normalized format."""
    await require_read_permission(http_request, FeatureGroup.WEBPROXY)
    try:
        service = get_session_vyos_service(http_request)
        full_config = await run_in_threadpool(service.get_full_config, refresh=refresh)

        wp = full_config.get("service", {}).get("webproxy", {})
        if not wp:
            return WebProxyConfig()

        return WebProxyConfig(
            append_domain=_parse_scalar(wp.get("append-domain")),
            cache_size=_parse_int(wp.get("cache-size")),
            default_port=_parse_int(wp.get("default-port")),
            maximum_object_size=_parse_int(wp.get("maximum-object-size")),
            mem_cache_size=_parse_int(wp.get("mem-cache-size")),
            minimum_object_size=_parse_int(wp.get("minimum-object-size")),
            outgoing_address=_parse_scalar(wp.get("outgoing-address")),
            reply_body_max_size=_parse_int(wp.get("reply-body-max-size")),
            disable_access_log="disable-access-log" in wp,
            domain_block=_parse_multi(wp.get("domain-block")),
            domain_noncache=_parse_multi(wp.get("domain-noncache")),
            reply_block_mime=_parse_multi(wp.get("reply-block-mime")),
            safe_ports=_parse_multi(wp.get("safe-ports")),
            ssl_safe_ports=_parse_multi(wp.get("ssl-safe-ports")),
            authentication=_parse_authentication(wp.get("authentication", {})),
            cache_peers=_parse_cache_peers(wp.get("cache-peer", {})),
            listen_addresses=_parse_listen_addresses(wp.get("listen-address", {})),
            url_filtering=_parse_url_filtering(wp.get("url-filtering", {})),
        )
    except Exception:
        logger.exception("Unhandled error in get_webproxy_config")
        raise HTTPException(status_code=500, detail="Internal server error")


# ============================================================================
# Endpoint 3: Batch Operations
# ============================================================================


@router.post("/batch", response_model=VyOSResponse)
async def webproxy_batch_configure(http_request: Request, body: WebProxyBatchRequest):
    """Execute a batch of webproxy configuration operations atomically."""
    await require_write_permission(http_request, FeatureGroup.WEBPROXY)
    try:
        service = get_session_vyos_service(http_request)
        builder = WebProxyBatchBuilder(version=service.get_version())

        for operation in body.operations:
            if operation.op in _INTERNAL_BUILDER_METHODS or operation.op.startswith("_"):
                raise HTTPException(
                    status_code=400,
                    detail=f"Operation not allowed: {operation.op}",
                )

            method = getattr(builder, operation.op)
            sig = inspect.signature(method)
            params = [p for p in sig.parameters.keys() if p != "self"]

            if len(params) == 0:
                method()
            elif len(params) == 1:
                if operation.value is not None:
                    method(operation.value)
                else:
                    method()
            elif len(params) >= 2:
                if operation.value and "," in operation.value:
                    parts = operation.value.split(",", len(params) - 1)
                    method(*parts)
                elif operation.value:
                    method(operation.value)

        response = service.execute_batch(builder)
        return VyOSResponse(
            success=response.status == 200,
            data={"message": "Web proxy configuration updated"},
            error=response.error if response.error else None,
        )
    except HTTPException:
        raise
    except AttributeError as e:
        raise HTTPException(status_code=400, detail=f"Unknown operation: {e}")
    except Exception:
        logger.exception("Unhandled error in webproxy_batch_configure")
        raise HTTPException(status_code=500, detail="Internal server error")


# ============================================================================
# Config parsers
# ============================================================================


def _parse_multi(value) -> List[str]:
    if value is None:
        return []
    if isinstance(value, list):
        return sorted(str(v) for v in value)
    if isinstance(value, dict):
        return sorted(value.keys())
    return [str(value)]


def _parse_int(value) -> Optional[int]:
    if isinstance(value, list):
        value = value[0] if value else None
    if value is None:
        return None
    try:
        return int(value)
    except (ValueError, TypeError):
        return None


def _parse_scalar(value) -> Optional[str]:
    if value is None:
        return None
    if isinstance(value, list):
        return str(value[0]) if value else None
    return str(value)


def _parse_authentication(raw: dict) -> WebProxyAuthentication:
    if not isinstance(raw, dict) or not raw:
        return WebProxyAuthentication()
    ldap_raw = raw.get("ldap", {}) if isinstance(raw.get("ldap"), dict) else {}
    return WebProxyAuthentication(
        children=_parse_int(raw.get("children")),
        credentials_ttl=_parse_int(raw.get("credentials-ttl")),
        method=_parse_scalar(raw.get("method")),
        realm=_parse_scalar(raw.get("realm")),
        ldap=WebProxyLdap(
            server=_parse_scalar(ldap_raw.get("server")),
            base_dn=_parse_scalar(ldap_raw.get("base-dn")),
            bind_dn=_parse_scalar(ldap_raw.get("bind-dn")),
            filter_expression=_parse_scalar(ldap_raw.get("filter-expression")),
            password=_parse_scalar(ldap_raw.get("password")),
            username_attribute=_parse_scalar(ldap_raw.get("username-attribute")),
            port=_parse_int(ldap_raw.get("port")),
            version=_parse_scalar(ldap_raw.get("version")),
            persistent_connection="persistent-connection" in ldap_raw,
            use_ssl="use-ssl" in ldap_raw,
        ),
    )


def _parse_cache_peers(raw: dict) -> List[CachePeer]:
    if not isinstance(raw, dict):
        return []
    peers = []
    for name, attrs in sorted(raw.items()):
        if not isinstance(attrs, dict):
            attrs = {}
        peers.append(CachePeer(
            name=name,
            address=_parse_scalar(attrs.get("address")),
            http_port=_parse_int(attrs.get("http-port")),
            icp_port=_parse_int(attrs.get("icp-port")),
            options=_parse_scalar(attrs.get("options")),
            type=_parse_scalar(attrs.get("type")),
        ))
    return peers


def _parse_listen_addresses(raw: dict) -> List[ListenAddress]:
    if not isinstance(raw, dict):
        return []
    addresses = []
    for addr, attrs in sorted(raw.items()):
        if not isinstance(attrs, dict):
            attrs = {}
        addresses.append(ListenAddress(
            address=addr,
            port=_parse_int(attrs.get("port")),
            disable_transparent="disable-transparent" in attrs,
        ))
    return addresses


def _parse_url_filtering(raw: dict) -> UrlFiltering:
    if not isinstance(raw, dict) or not raw:
        return UrlFiltering()
    return UrlFiltering(
        disable="disable" in raw,
        squidguard=_parse_squidguard(raw.get("squidguard", {})),
    )


def _parse_squidguard(raw: dict) -> SquidGuard:
    if not isinstance(raw, dict) or not raw:
        return SquidGuard()
    auto_update = raw.get("auto-update", {})
    auto_update_hour = None
    if isinstance(auto_update, dict):
        auto_update_hour = _parse_int(auto_update.get("update-hour"))
    return SquidGuard(
        allow_categories=_parse_multi(raw.get("allow-category")),
        block_categories=_parse_multi(raw.get("block-category")),
        log=_parse_multi(raw.get("log")),
        local_block=_parse_multi(raw.get("local-block")),
        local_block_keyword=_parse_multi(raw.get("local-block-keyword")),
        local_block_url=_parse_multi(raw.get("local-block-url")),
        local_ok=_parse_multi(raw.get("local-ok")),
        local_ok_url=_parse_multi(raw.get("local-ok-url")),
        allow_ipaddr_url="allow-ipaddr-url" in raw,
        enable_safe_search="enable-safe-search" in raw,
        default_action=_parse_scalar(raw.get("default-action")),
        redirect_url=_parse_scalar(raw.get("redirect-url")),
        auto_update_hour=auto_update_hour,
        rules=_parse_squidguard_rules(raw.get("rule", {})),
        source_groups=_parse_source_groups(raw.get("source-group", {})),
        time_periods=_parse_time_periods(raw.get("time-period", {})),
    )


def _parse_squidguard_rules(raw: dict) -> List[SquidGuardRule]:
    if not isinstance(raw, dict):
        return []
    rules = []
    for number, attrs in sorted(raw.items(), key=lambda kv: _parse_int(kv[0]) or 0):
        if not isinstance(attrs, dict):
            attrs = {}
        rules.append(SquidGuardRule(
            number=str(number),
            allow_categories=_parse_multi(attrs.get("allow-category")),
            block_categories=_parse_multi(attrs.get("block-category")),
            log=_parse_multi(attrs.get("log")),
            local_block=_parse_multi(attrs.get("local-block")),
            local_block_keyword=_parse_multi(attrs.get("local-block-keyword")),
            local_block_url=_parse_multi(attrs.get("local-block-url")),
            local_ok=_parse_multi(attrs.get("local-ok")),
            local_ok_url=_parse_multi(attrs.get("local-ok-url")),
            allow_ipaddr_url="allow-ipaddr-url" in attrs,
            enable_safe_search="enable-safe-search" in attrs,
            default_action=_parse_scalar(attrs.get("default-action")),
            redirect_url=_parse_scalar(attrs.get("redirect-url")),
            source_group=_parse_scalar(attrs.get("source-group")),
            time_period=_parse_scalar(attrs.get("time-period")),
        ))
    return rules


def _parse_source_groups(raw: dict) -> List[SquidGuardSourceGroup]:
    if not isinstance(raw, dict):
        return []
    groups = []
    for name, attrs in sorted(raw.items()):
        if not isinstance(attrs, dict):
            attrs = {}
        groups.append(SquidGuardSourceGroup(
            name=name,
            address=_parse_multi(attrs.get("address")),
            domain=_parse_multi(attrs.get("domain")),
            ldap_ip_search=_parse_multi(attrs.get("ldap-ip-search")),
            ldap_user_search=_parse_multi(attrs.get("ldap-user-search")),
            description=_parse_scalar(attrs.get("description")),
            user=_parse_scalar(attrs.get("user")),
        ))
    return groups


def _parse_time_periods(raw: dict) -> List[SquidGuardTimePeriod]:
    if not isinstance(raw, dict):
        return []
    periods = []
    for name, attrs in sorted(raw.items()):
        if not isinstance(attrs, dict):
            attrs = {}
        days_raw = attrs.get("days", {})
        days = []
        if isinstance(days_raw, dict):
            for day, day_attrs in sorted(days_raw.items()):
                time_val = None
                if isinstance(day_attrs, dict):
                    time_val = _parse_scalar(day_attrs.get("time"))
                days.append(TimePeriodDay(day=day, time=time_val))
        periods.append(SquidGuardTimePeriod(
            name=name,
            description=_parse_scalar(attrs.get("description")),
            days=days,
        ))
    return periods
